### PR TITLE
Fix: `prefer-const` false positive reading before writing

### LINKED ISSUE #5074: prefer-const rule false positive
Version 2.0.0.beta.2 also checked with latest master https://github.com/eslint/eslint/commit/6433b7b4372468d3470d32f6c2285fa9f06e24fa.
Code is:
```js
let id;

function foo() {
    if (typeof id !== 'undefined') {
        return;
    }
    id = setInterval(() => {
        // some stuff
    }, 250);
}

foo();
```
The output is `1:5  error  'id' is never modified, use 'const' instead  prefer-const`.
But `id` is modified and declaration with `let` shouldn't alert error in this case.
(moved from #5024 as it is not the same)

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30216765-prefer-const-rule-false-positive?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mysticatea**
Fixes #5074.

### CODE REVIEW

**Comment by mysticatea** [comment]
Wow, cool.

File: tests/lib/rules/prefer-const.js [hunk 1]
```diff
@@ -133,6 +153,16 @@ ruleTester.run("prefer-const", rule, {
             code: "(function() { let x; { x = 0; foo(x); } })();",
             parserOptions: { ecmaVersion: 6 },
             errors: [{ message: "'x' is never modified, use 'const' instead.", type: "Identifier"}]
+        },
+        {
+            code: "let x; for (const a of [1,2,3]) { x = foo(); bar(x); }",
+            parserOptions: { ecmaVersion: 6 },
```
Comment (nzakas) [inline tests/lib/rules/prefer-const.js]: I'm not sure this is correct. `x = foo()` could return a different value each time it's called. I don't think you can safely replace this with a const.

Comment (mysticatea) [inline tests/lib/rules/prefer-const.js]: This is good working.

```js
let x; for (const a of [1,2,3]) { x = 2 * a; bar(x); }
for (const a of [1,2,3]) { const x = 2 * a; bar(x); }
```

Comment (nzakas) [inline tests/lib/rules/prefer-const.js]: Ah! I see now, thanks! :+1:

**Comment by nzakas** [comment]
Lgtm

**Pull request merged**